File: chatguard-web/src/lib/storage/encryptedFileStorage.ts
/**
 * Encrypted File Storage using Shield.
 *
 * All files are stored encrypted at rest using Shield's StreamCipher-equivalent.
 * Files can be downloaded either encrypted (for backup/transfer) or decrypted (for viewing).
 *
 * File format:
 * - Magic header: "SHLD_ENC" (8 bytes)
 * - Version: 0x01 (1 byte)
 * - Metadata length: 4 bytes (little-endian)
 * - Encrypted metadata: JSON with filename, type, size, checksum
 * - Encrypted content: AES-GCM encrypted data
 */

import { Shield } from '../shield/crypto';

const MAGIC_HEADER = new TextEncoder().encode('SHLD_ENC');
const VERSION = 0x01;
const HEADER_SIZE = 8 + 1 + 4; // magic + version + metadata_length
const CHUNK_SIZE = 65536; // 64KB chunks for streaming

/**
 * File metadata stored with encrypted file.
 */
export interface FileMetadata {
  originalName: string;
  mimeType: string;
  originalSize: number;
  createdAt: number;
  checksum?: string; // SHA256 of plaintext
}

/**
 * Reference to an encrypted file.
 */
export interface EncryptedFileRef {
  id: string;
  contactId: string;
  encryptedSize: number;
  metadata?: FileMetadata;
}

/**
 * Encrypted file storage using Shield and IndexedDB.
 */
export class EncryptedFileStorage {
  private dbPromise: Promise<IDBDatabase> | null = null;

  constructor(private keyStore: KeyStore) {}

  /**
   * Save a file encrypted with Shield.
   */
  async saveEncrypted(
    contactId: string,
    file: File,
    metadata?: Partial<FileMetadata>
  ): Promise<EncryptedFileRef> {
    const fileId = crypto.randomUUID();

    // Read file content
    const plaintext = new Uint8Array(await file.arrayBuffer());

    // Calculate checksum
    const checksum = await this.calculateChecksum(plaintext);

    // Build metadata
    const fullMetadata: FileMetadata = {
      originalName: metadata?.originalName ?? file.name,
      mimeType: (metadata?.mimeType ?? file.type) || 'application/octet-stream',
      originalSize: plaintext.length,
      createdAt: metadata?.createdAt ?? Date.now(),
      checksum,
    };

    // Get file key
    const fileKey = await this.deriveFileKey(contactId, fileId);

    // Encrypt metadata
    const metadataJson = JSON.stringify(fullMetadata);
    const metadataBytes = new TextEncoder().encode(metadataJson);
    const encryptedMetadata = await Shield.quickEncrypt(fileKey, metadataBytes);

    // Encrypt content
    const encryptedContent = await Shield.quickEncrypt(fileKey, plaintext);

    // Build encrypted file
    const encryptedFile = this.buildEncryptedFile(encryptedMetadata, encryptedContent);

    // Store in IndexedDB
    const db = await this.openDatabase();
    await this.storeFile(db, fileId, contactId, encryptedFile);

    return {
      id: fileId,
      contactId,
      encryptedSize: encryptedFile.length,
      metadata: fullMetadata,
    };
  }

  /**
   * Save from Uint8Array (for received files).
   */
  async saveEncryptedFromBytes(
    contactId: string,
    data: Uint8Array,
    metadata: FileMetadata
  ): Promise<EncryptedFileRef> {
    const fileId = crypto.randomUUID();

    // Calculate checksum
    const checksum = await this.calculateChecksum(data);
    const fullMetadata = { ...metadata, checksum };

    // Get file key
    const fileKey = await this.deriveFileKey(contactId, fileId);

    // Encrypt metadata
    const metadataJson = JSON.stringify(fullMetadata);
    const metadataBytes = new TextEncoder().encode(metadataJson);
    const encryptedMetadata = await Shield.quickEncrypt(fileKey, metadataBytes);

    // Encrypt content
    const encryptedContent = await Shield.quickEncrypt(fileKey, data);

    // Build encrypted file
    const encryptedFile = this.buildEncryptedFile(encryptedMetadata, encryptedContent);

    // Store in IndexedDB
    const db = await this.openDatabase();
    await this.storeFile(db, fileId, contactId, encryptedFile);

    return {
      id: fileId,
      contactId,
      encryptedSize: encryptedFile.length,
      metadata: fullMetadata,
    };
  }

  /**
   * Load and decrypt a file.
   */
  async loadDecrypted(fileRef: EncryptedFileRef): Promise<Blob> {
    const db = await this.openDatabase();
    const encryptedFile = await this.retrieveFile(db, fileRef.id);

    if (!encryptedFile) {
      throw new Error(`File not found: ${fileRef.id}`);
    }

    // Parse header and decrypt
    const { metadata, content } = await this.parseAndDecrypt(
      encryptedFile,
      fileRef.contactId,
      fileRef.id
    );

    // Verify checksum
    if (metadata.checksum) {
      const actualChecksum = await this.calculateChecksum(content);
      if (actualChecksum !== metadata.checksum) {
        throw new Error('Checksum mismatch: file may be corrupted');
      }
    }

    return new Blob([content], { type: metadata.mimeType });
  }

  /**
   * Download file in encrypted form (for backup/transfer).
   * The file remains encrypted and can be transferred to another device.
   */
  async downloadEncrypted(fileRef: EncryptedFileRef): Promise<Blob> {
    const db = await this.openDatabase();
    const encryptedFile = await this.retrieveFile(db, fileRef.id);

    if (!encryptedFile) {
      throw new Error(`File not found: ${fileRef.id}`);
    }

    return new Blob([encryptedFile], { type: 'application/octet-stream' });
  }

  /**
   * Download file in decrypted form (for viewing/sharing).
   */
  async downloadDecrypted(fileRef: EncryptedFileRef): Promise<Blob> {
    return this.loadDecrypted(fileRef);
  }

  /**
   * Get a download URL for encrypted file.
   * Creates a temporary object URL.
   */
  async getEncryptedDownloadUrl(fileRef: EncryptedFileRef): Promise<string> {
    const blob = await this.downloadEncrypted(fileRef);
    return URL.createObjectURL(blob);
  }

  /**
   * Get a download URL for decrypted file.
   * Creates a temporary object URL.
   */
  async getDecryptedDownloadUrl(fileRef: EncryptedFileRef): Promise<string> {
    const blob = await this.downloadDecrypted(fileRef);
    return URL.createObjectURL(blob);
  }

  /**
   * Trigger browser download of encrypted file.
   */
  async triggerEncryptedDownload(fileRef: EncryptedFileRef, filename?: string): Promise<void> {
    const url = await this.getEncryptedDownloadUrl(fileRef);
    const metadata = await this.getMetadata(fileRef);
    const downloadName = filename ?? `${metadata?.originalName ?? fileRef.id}.enc`;

    this.triggerDownload(url, downloadName);
    setTimeout(() => URL.revokeObjectURL(url), 60000);
  }

  /**
   * Trigger browser download of decrypted file.
   */
  async triggerDecryptedDownload(fileRef: EncryptedFileRef, filename?: string): Promise<void> {
    const url = await this.getDecryptedDownloadUrl(fileRef);
    const metadata = await this.getMetadata(fileRef);
    const downloadName = filename ?? metadata?.originalName ?? fileRef.id;

    this.triggerDownload(url, downloadName);
    setTimeout(() => URL.revokeObjectURL(url), 60000);
  }

  /**
   * Get metadata for an encrypted file without decrypting content.
   */
  async getMetadata(fileRef: EncryptedFileRef): Promise<FileMetadata | null> {
    const db = await this.openDatabase();
    const encryptedFile = await this.retrieveFile(db, fileRef.id);

    if (!encryptedFile) {
      return null;
    }

    try {
      const { metadata } = await this.parseAndDecrypt(
        encryptedFile,
        fileRef.contactId,
        fileRef.id,
        true // metadata only
      );
      return metadata;
    } catch {
      return null;
    }
  }

  /**
   * Import an encrypted file received from another device.
   */
  async importEncrypted(
    contactId: string,
    encryptedData: Uint8Array,
    fileId?: string
  ): Promise<EncryptedFileRef> {
    const id = fileId ?? crypto.randomUUID();

    // Validate header
    if (encryptedData.length < HEADER_SIZE) {
      throw new Error('Invalid encrypted file: too small');
    }

    const magic = encryptedData.slice(0, 8);
    if (!this.arraysEqual(magic, MAGIC_HEADER)) {
      throw new Error('Invalid encrypted file: bad magic header');
    }

    // Store in IndexedDB
    const db = await this.openDatabase();
    await this.storeFile(db, id, contactId, encryptedData);

    return {
      id,
      contactId,
      encryptedSize: encryptedData.length,
    };
  }

  /**
   * Delete an encrypted file.
   */
  async delete(fileRef: EncryptedFileRef): Promise<void> {
    const db = await this.openDatabase();
    await this.deleteFile(db, fileRef.id);
  }

  /**
   * List all encrypted files for a contact.
   */
  async listFiles(contactId: string): Promise<EncryptedFileRef[]> {
    const db = await this.openDatabase();
    return new Promise((resolve, reject) => {
      const tx = db.transaction('files', 'readonly');
      const store = tx.objectStore('files');
      const index = store.index('contactId');
      const request = index.getAll(contactId);

      request.onsuccess = () => {
        const files = request.result.map((record: StoredFile) => ({
          id: record.id,
          contactId: record.contactId,
          encryptedSize: record.data.length,
        }));
        resolve(files);
      };

      request.onerror = () => reject(request.error);
    });
  }

  /**
   * Get total storage used by encrypted files.
   */
  async getStorageUsed(): Promise<number> {
    const db = await this.openDatabase();
    return new Promise((resolve, reject) => {
      const tx = db.transaction('files', 'readonly');
      const store = tx.objectStore('files');
      const request = store.getAll();

      request.onsuccess = () => {
        const total = request.result.reduce(
          (sum: number, record: StoredFile) => sum + record.data.length,
          0
        );
        resolve(total);
      };

      request.onerror = () => reject(request.error);
    });
  }

  // Private helpers

  private async deriveFileKey(contactId: string, fileId: string): Promise<Uint8Array> {
    const mediaKey = await this.keyStore.getKey(`media_key_${contactId}`);
    if (!mediaKey) {
      throw new Error(`No media key for contact: ${contactId}`);
    }

    // HKDF-like derivation: HMAC(mediaKey, fileId)
    const cryptoKey = await crypto.subtle.importKey(
      'raw',
      mediaKey,
      { name: 'HMAC', hash: 'SHA-256' },
      false,
      ['sign']
    );

    const fileIdBytes = new TextEncoder().encode(fileId);
    const derived = await crypto.subtle.sign('HMAC', cryptoKey, fileIdBytes);
    return new Uint8Array(derived);
  }

  private buildEncryptedFile(
    encryptedMetadata: Uint8Array,
    encryptedContent: Uint8Array
  ): Uint8Array {
    const metadataLength = this.intToLittleEndian(encryptedMetadata.length);

    const totalSize =
      MAGIC_HEADER.length + 1 + 4 + encryptedMetadata.length + encryptedContent.length;
    const result = new Uint8Array(totalSize);

    let offset = 0;

    // Magic header
    result.set(MAGIC_HEADER, offset);
    offset += MAGIC_HEADER.length;

    // Version
    result[offset] = VERSION;
    offset += 1;

    // Metadata length
    result.set(metadataLength, offset);
    offset += 4;

    // Encrypted metadata
    result.set(encryptedMetadata, offset);
    offset += encryptedMetadata.length;

    // Encrypted content
    result.set(encryptedContent, offset);

    return result;
  }

  private async parseAndDecrypt(
    encryptedFile: Uint8Array,
    contactId: string,
    fileId: string,
    metadataOnly = false
  ): Promise<{ metadata: FileMetadata; content: Uint8Array }> {
    // Validate magic header
    const magic = encryptedFile.slice(0, 8);
    if (!this.arraysEqual(magic, MAGIC_HEADER)) {
      throw new Error('Invalid file: bad magic header');
    }

    // Read version
    const version = encryptedFile[8];
    if (version !== VERSION) {
      throw new Error(`Unsupported file version: ${version}`);
    }

    // Read metadata length
    const metadataLength = this.littleEndianToInt(encryptedFile.slice(9, 13));

    // Read encrypted metadata
    const encryptedMetadata = encryptedFile.slice(13, 13 + metadataLength);

    // Get file key
    const fileKey = await this.deriveFileKey(contactId, fileId);

    // Decrypt metadata
    const metadataBytes = await Shield.quickDecrypt(fileKey, encryptedMetadata);
    const metadataJson = new TextDecoder().decode(metadataBytes);
    const metadata: FileMetadata = JSON.parse(metadataJson);

    if (metadataOnly) {
      return { metadata, content: new Uint8Array(0) };
    }

    // Decrypt content
    const encryptedContent = encryptedFile.slice(13 + metadataLength);
    const content = await Shield.quickDecrypt(fileKey, encryptedContent);

    return { metadata, content };
  }

  private async calculateChecksum(data: Uint8Array): Promise<string> {
    const hash = await crypto.subtle.digest('SHA-256', data);
    return Array.from(new Uint8Array(hash))
      .map((b) => b.toString(16).padStart(2, '0'))
      .join('');
  }

  private intToLittleEndian(value: number): Uint8Array {
    return new Uint8Array([
      value & 0xff,
      (value >> 8) & 0xff,
      (value >> 16) & 0xff,
      (value >> 24) & 0xff,
    ]);
  }

  private littleEndianToInt(bytes: Uint8Array): number {
    return bytes[0] | (bytes[1] << 8) | (bytes[2] << 16) | (bytes[3] << 24);
  }

  private arraysEqual(a: Uint8Array, b: Uint8Array): boolean {
    if (a.length !== b.length) return false;
    for (let i = 0; i < a.length; i++) {
      if (a[i] !== b[i]) return false;
    }
    return true;
  }

  private triggerDownload(url: string, filename: string): void {
    const a = document.createElement('a');
    a.href = url;
    a.download = filename;
    document.body.appendChild(a);
    a.click();
    document.body.removeChild(a);
  }

  // IndexedDB operations

  private async openDatabase(): Promise<IDBDatabase> {
    if (this.dbPromise) {
      return this.dbPromise;
    }

    this.dbPromise = new Promise((resolve, reject) => {
      const request = indexedDB.open('chatguard-files', 1);

      request.onerror = () => reject(request.error);
      request.onsuccess = () => resolve(request.result);

      request.onupgradeneeded = (event) => {
        const db = (event.target as IDBOpenDBRequest).result;

        if (!db.objectStoreNames.contains('files')) {
          const store = db.createObjectStore('files', { keyPath: 'id' });
          store.createIndex('contactId', 'contactId', { unique: false });
        }
      };
    });

    return this.dbPromise;
  }

  private async storeFile(
    db: IDBDatabase,
    fileId: string,
    contactId: string,
    data: Uint8Array
  ): Promise<void> {
    return new Promise((resolve, reject) => {
      const tx = db.transaction('files', 'readwrite');
      const store = tx.objectStore('files');

      const record: StoredFile = {
        id: fileId,
        contactId,
        data: Array.from(data), // Convert to array for IndexedDB
        createdAt: Date.now(),
      };

      const request = store.put(record);
      request.onsuccess = () => resolve();
      request.onerror = () => reject(request.error);
    });
  }

  private async retrieveFile(db: IDBDatabase, fileId: string): Promise<Uint8Array | null> {
    return new Promise((resolve, reject) => {
      const tx = db.transaction('files', 'readonly');
      const store = tx.objectStore('files');
      const request = store.get(fileId);

      request.onsuccess = () => {
        const record = request.result as StoredFile | undefined;
        if (record) {
          resolve(new Uint8Array(record.data));
        } else {
          resolve(null);
        }
      };

      request.onerror = () => reject(request.error);
    });
  }

  private async deleteFile(db: IDBDatabase, fileId: string): Promise<void> {
    return new Promise((resolve, reject) => {
      const tx = db.transaction('files', 'readwrite');
      const store = tx.objectStore('files');
      const request = store.delete(fileId);

      request.onsuccess = () => resolve();
      request.onerror = () => reject(request.error);
    });
  }
}

interface StoredFile {
  id: string;
  contactId: string;
  data: number[];
  createdAt: number;
}

/**
 * Key store interface for accessing encryption keys.
 */
export interface KeyStore {
  getKey(keyId: string): Promise<Uint8Array | null>;
}

/**
 * Create EncryptedFileStorage with the Shield crypto key store.
 */
export function createEncryptedFileStorage(keyStore: KeyStore): EncryptedFileStorage {
  return new EncryptedFileStorage(keyStore);
}

File: chatguard-android/app/src/main/java/ai/guard8/chatguard/storage/EncryptedFileStorage.kt
package ai.guard8.chatguard.storage

import android.content.Context
import ai.guard8.chatguard.crypto.KeyManager
import ai.guard8.shield.Shield
import ai.guard8.shield.StreamCipher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.InputStream
import java.io.OutputStream
import java.util.UUID

/**
 * Encrypted file metadata.
 */
@Serializable
data class FileMetadata(
    val originalName: String,
    val mimeType: String,
    val originalSize: Long,
    val createdAt: Long = System.currentTimeMillis(),
    val checksum: String? = null  // SHA256 of plaintext
)

/**
 * Reference to an encrypted file.
 */
data class EncryptedFile(
    val id: String,
    val path: String,
    val encryptedSize: Long,
    val contactId: String
)

/**
 * Encrypted file storage using Shield StreamCipher.
 *
 * File format:
 * - Magic header: "SHLD_ENC" (8 bytes)
 * - Version: 0x01 (1 byte)
 * - Metadata length: 4 bytes (little-endian)
 * - Encrypted metadata: JSON with filename, type, size, checksum
 * - Encrypted content: StreamCipher encrypted data
 *
 * All files are stored encrypted at rest. Files can be downloaded
 * either encrypted (for backup/transfer) or decrypted (for viewing).
 */
class EncryptedFileStorage(
    private val context: Context,
    private val keyManager: KeyManager
) {
    companion object {
        private val MAGIC_HEADER = "SHLD_ENC".toByteArray(Charsets.US_ASCII)
        private const val VERSION: Byte = 0x01
        private const val HEADER_SIZE = 8 + 1 + 4  // magic + version + metadata_length

        private val json = Json { ignoreUnknownKeys = true }
    }

    private val storageDir: File by lazy {
        File(context.filesDir, "encrypted_files").also { it.mkdirs() }
    }

    /**
     * Save a file encrypted with Shield.
     *
     * @param contactId Contact ID for key derivation
     * @param inputFile Source file to encrypt
     * @param metadata File metadata (name, type, etc.)
     * @return Reference to the encrypted file
     */
    suspend fun saveEncrypted(
        contactId: String,
        inputFile: File,
        metadata: FileMetadata
    ): EncryptedFile = withContext(Dispatchers.IO) {
        val fileId = UUID.randomUUID().toString()
        val outputFile = File(storageDir, "$fileId.enc")

        // Calculate checksum of plaintext
        val checksum = calculateChecksum(inputFile)
        val metadataWithChecksum = metadata.copy(checksum = checksum)

        // Get or derive file key
        val fileKey = deriveFileKey(contactId, fileId)

        // Encrypt metadata
        val metadataJson = json.encodeToString(metadataWithChecksum)
        val encryptedMetadata = Shield.quickEncrypt(fileKey, metadataJson.toByteArray(Charsets.UTF_8))

        // Write encrypted file
        FileOutputStream(outputFile).use { fos ->
            // Write header
            fos.write(MAGIC_HEADER)
            fos.write(VERSION.toInt())
            fos.write(intToLittleEndian(encryptedMetadata.size))

            // Write encrypted metadata
            fos.write(encryptedMetadata)

            // Encrypt and write content using StreamCipher
            StreamCipher.create(fileKey).use { cipher ->
                FileInputStream(inputFile).use { fis ->
                    cipher.encryptStream(fis, fos)
                }
            }
        }

        EncryptedFile(
            id = fileId,
            path = outputFile.absolutePath,
            encryptedSize = outputFile.length(),
            contactId = contactId
        )
    }

    /**
     * Save from InputStream (for received files).
     */
    suspend fun saveEncrypted(
        contactId: String,
        inputStream: InputStream,
        metadata: FileMetadata
    ): EncryptedFile = withContext(Dispatchers.IO) {
        // Write to temp file first to calculate checksum
        val tempFile = File.createTempFile("incoming_", ".tmp", context.cacheDir)
        try {
            tempFile.outputStream().use { fos ->
                inputStream.copyTo(fos)
            }
            saveEncrypted(contactId, tempFile, metadata)
        } finally {
            tempFile.delete()
        }
    }

    /**
     * Load and decrypt a file.
     *
     * @param encryptedFile Reference to the encrypted file
     * @return Decrypted file in cache directory
     */
    suspend fun loadDecrypted(
        encryptedFile: EncryptedFile
    ): File = withContext(Dispatchers.IO) {
        val inputFile = File(encryptedFile.path)
        require(inputFile.exists()) { "Encrypted file not found: ${encryptedFile.path}" }

        // Read and parse header
        val (metadata, contentOffset) = readHeader(inputFile, encryptedFile.contactId, encryptedFile.id)

        // Create output file with original extension
        val extension = metadata.originalName.substringAfterLast('.', "")
        val outputFile = File(context.cacheDir, "${encryptedFile.id}.$extension")

        // Get file key
        val fileKey = deriveFileKey(encryptedFile.contactId, encryptedFile.id)

        // Decrypt content
        FileInputStream(inputFile).use { fis ->
            fis.skip(contentOffset.toLong())

            StreamCipher.create(fileKey).use { cipher ->
                FileOutputStream(outputFile).use { fos ->
                    cipher.decryptStream(fis, fos)
                }
            }
        }

        // Verify checksum if present
        metadata.checksum?.let { expectedChecksum ->
            val actualChecksum = calculateChecksum(outputFile)
            require(actualChecksum == expectedChecksum) {
                "Checksum mismatch: file may be corrupted"
            }
        }

        outputFile
    }

    /**
     * Download file in encrypted form (for backup/transfer).
     * The file remains encrypted and can be transferred to another device.
     *
     * @param encryptedFile Reference to the encrypted file
     * @return Raw encrypted bytes including header
     */
    suspend fun downloadEncrypted(
        encryptedFile: EncryptedFile
    ): ByteArray = withContext(Dispatchers.IO) {
        val file = File(encryptedFile.path)
        require(file.exists()) { "Encrypted file not found: ${encryptedFile.path}" }
        file.readBytes()
    }

    /**
     * Download file in decrypted form (for viewing/sharing).
     *
     * @param encryptedFile Reference to the encrypted file
     * @return Decrypted file bytes
     */
    suspend fun downloadDecrypted(
        encryptedFile: EncryptedFile
    ): ByteArray = withContext(Dispatchers.IO) {
        val decryptedFile = loadDecrypted(encryptedFile)
        try {
            decryptedFile.readBytes()
        } finally {
            decryptedFile.delete()
        }
    }

    /**
     * Stream decrypted content to an OutputStream.
     * Useful for sharing or displaying large files.
     */
    suspend fun streamDecrypted(
        encryptedFile: EncryptedFile,
        outputStream: OutputStream
    ) = withContext(Dispatchers.IO) {
        val inputFile = File(encryptedFile.path)
        require(inputFile.exists()) { "Encrypted file not found: ${encryptedFile.path}" }

        val (_, contentOffset) = readHeader(inputFile, encryptedFile.contactId, encryptedFile.id)
        val fileKey = deriveFileKey(encryptedFile.contactId, encryptedFile.id)

        FileInputStream(inputFile).use { fis ->
            fis.skip(contentOffset.toLong())

            StreamCipher.create(fileKey).use { cipher ->
                cipher.decryptStream(fis, outputStream)
            }
        }
    }

    /**
     * Get metadata for an encrypted file without decrypting content.
     */
    suspend fun getMetadata(
        encryptedFile: EncryptedFile
    ): FileMetadata = withContext(Dispatchers.IO) {
        val inputFile = File(encryptedFile.path)
        require(inputFile.exists()) { "Encrypted file not found: ${encryptedFile.path}" }
        readHeader(inputFile, encryptedFile.contactId, encryptedFile.id).first
    }

    /**
     * Import an encrypted file received from another device.
     * The file must have been encrypted with the same contact's key.
     */
    suspend fun importEncrypted(
        contactId: String,
        encryptedData: ByteArray,
        fileId: String = UUID.randomUUID().toString()
    ): EncryptedFile = withContext(Dispatchers.IO) {
        // Validate header
        require(encryptedData.size > HEADER_SIZE) { "Invalid encrypted file: too small" }
        require(encryptedData.sliceArray(0 until 8).contentEquals(MAGIC_HEADER)) {
            "Invalid encrypted file: bad magic header"
        }

        val outputFile = File(storageDir, "$fileId.enc")
        outputFile.writeBytes(encryptedData)

        EncryptedFile(
            id = fileId,
            path = outputFile.absolutePath,
            encryptedSize = outputFile.length(),
            contactId = contactId
        )
    }

    /**
     * Delete an encrypted file.
     */
    suspend fun delete(encryptedFile: EncryptedFile) = withContext(Dispatchers.IO) {
        File(encryptedFile.path).delete()
    }

    /**
     * List all encrypted files for a contact.
     */
    suspend fun listFiles(contactId: String): List<EncryptedFile> = withContext(Dispatchers.IO) {
        storageDir.listFiles { file -> file.extension == "enc" }
            ?.mapNotNull { file ->
                try {
                    val id = file.nameWithoutExtension
                    // Try to read header to verify it's for this contact
                    val header = readHeaderRaw(file)
                    if (header != null) {
                        EncryptedFile(
                            id = id,
                            path = file.absolutePath,
                            encryptedSize = file.length(),
                            contactId = contactId
                        )
                    } else null
                } catch (_: Exception) {
                    null
                }
            }
            ?: emptyList()
    }

    // Private helpers

    private fun deriveFileKey(contactId: String, fileId: String): ByteArray {
        val mediaKey = keyManager.retrieveKey("media_key_$contactId")
            ?: throw IllegalStateException("No media key for contact: $contactId")

        // HKDF-like derivation using Shield: HMAC(mediaKey, fileId)
        return Shield.hmacSha256(mediaKey, fileId.toByteArray(Charsets.UTF_8))
    }

    private fun readHeader(
        file: File,
        contactId: String,
        fileId: String
    ): Pair<FileMetadata, Int> {
        FileInputStream(file).use { fis ->
            // Read magic header
            val magic = ByteArray(8)
            fis.read(magic)
            require(magic.contentEquals(MAGIC_HEADER)) { "Invalid file: bad magic header" }

            // Read version
            val version = fis.read().toByte()
            require(version == VERSION) { "Unsupported file version: $version" }

            // Read metadata length
            val lengthBytes = ByteArray(4)
            fis.read(lengthBytes)
            val metadataLength = littleEndianToInt(lengthBytes)

            // Read encrypted metadata
            val encryptedMetadata = ByteArray(metadataLength)
            fis.read(encryptedMetadata)

            // Decrypt metadata
            val fileKey = deriveFileKey(contactId, fileId)
            val metadataJson = Shield.quickDecrypt(fileKey, encryptedMetadata)
            val metadata = json.decodeFromString<FileMetadata>(String(metadataJson, Charsets.UTF_8))

            val contentOffset = HEADER_SIZE + metadataLength
            return Pair(metadata, contentOffset)
        }
    }

    private fun readHeaderRaw(file: File): ByteArray? {
        return try {
            FileInputStream(file).use { fis ->
                val magic = ByteArray(8)
                if (fis.read(magic) != 8) return null
                if (!magic.contentEquals(MAGIC_HEADER)) return null
                magic
            }
        } catch (_: Exception) {
            null
        }
    }

    private fun calculateChecksum(file: File): String {
        // Use Shield for SHA-256 hashing - read file and hash with Shield
        val fileBytes = file.readBytes()
        val hash = Shield.sha256(fileBytes)
        return hash.joinToString("") { "%02x".format(it) }
    }

    private fun intToLittleEndian(value: Int): ByteArray {
        return byteArrayOf(
            (value and 0xFF).toByte(),
            ((value shr 8) and 0xFF).toByte(),
            ((value shr 16) and 0xFF).toByte(),
            ((value shr 24) and 0xFF).toByte()
        )
    }

    private fun littleEndianToInt(bytes: ByteArray): Int {
        return (bytes[0].toInt() and 0xFF) or
                ((bytes[1].toInt() and 0xFF) shl 8) or
                ((bytes[2].toInt() and 0xFF) shl 16) or
                ((bytes[3].toInt() and 0xFF) shl 24)
    }
}

/**
 * Extension to create StreamCipher that encrypts from InputStream to OutputStream.
 */
private fun StreamCipher.encryptStream(input: InputStream, output: OutputStream) {
    val buffer = ByteArray(65536)  // 64KB chunks
    var bytesRead: Int
    while (input.read(buffer).also { bytesRead = it } != -1) {
        val chunk = if (bytesRead == buffer.size) buffer else buffer.copyOf(bytesRead)
        val encrypted = this.encryptChunk(chunk)
        output.write(encrypted)
    }
    // Write final block
    val finalBlock = this.finishEncrypt()
    if (finalBlock.isNotEmpty()) {
        output.write(finalBlock)
    }
}

/**
 * Extension to create StreamCipher that decrypts from InputStream to OutputStream.
 */
private fun StreamCipher.decryptStream(input: InputStream, output: OutputStream) {
    val buffer = ByteArray(65536 + 16)  // 64KB + auth tag
    var bytesRead: Int
    while (input.read(buffer).also { bytesRead = it } != -1) {
        val chunk = if (bytesRead == buffer.size) buffer else buffer.copyOf(bytesRead)
        val decrypted = this.decryptChunk(chunk)
        output.write(decrypted)
    }
    // Verify and write final block
    val finalBlock = this.finishDecrypt()
    if (finalBlock.isNotEmpty()) {
        output.write(finalBlock)
    }
}
